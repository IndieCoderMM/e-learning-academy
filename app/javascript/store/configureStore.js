import { configureStore } from '@reduxjs/toolkit';
import userReducer from './user';
import reservationReducer from './reservation';

export default configureStore({
  reducer: {
    user: userReducer,
    reservations: reservationReducer,
  },
});
