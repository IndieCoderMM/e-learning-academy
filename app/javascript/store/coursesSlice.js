/* eslint-disable no-param-reassign */
import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';

const initialState = {
  courses: [], loading: false,
};

const FETCHED_COURSES = 'courses/fetchCourses';
const apiEndpoint = '/api/courses';

export const fetchCourses = createAsyncThunk(
  FETCHED_COURSES,
  async () => {
    try {
      const response = await axios.get(apiEndpoint);
      return response.data;
    } catch (error) {
      throw Error('Error fetching courses');
    }
  },
);

const coursesSlice = createSlice({
  name: 'courses',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(fetchCourses.pending, (state) => {
        state.loading = true;
        state.error = null;
      })
      .addCase(fetchCourses.fulfilled, (state, action) => {
        state.loading = false;
        state.courses = action.payload;
      })
      .addCase(fetchCourses.rejected, (state, action) => {
        state.loading = false;
        state.error = action.error.message;
      });
  },
});

export const coursesActions = {
  fetchCourses,
};

export default coursesSlice.reducer;
