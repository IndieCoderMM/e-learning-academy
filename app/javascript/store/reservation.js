import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

export const getUserReservations = createAsyncThunk(
  'reservations/getUserReservations',
  async (userId) => {
    const response = await fetch(`/users/${userId}/reservations`);
    const data = await response.json();

    if (!response.ok) {
      throw new Error(data.message);
    }

    console.log(data);

    return data;
  },
);

const initialState = {
  data: [],
  status: 'idle',
  error: null,
};

const reservationSlice = createSlice({
  name: 'reservations',
  initialState,
  extraReducers(builder) {
    builder
      .addCase(getUserReservations.pending, (state) => {
        return { ...state, status: 'loading', error: null };
      })
      .addCase(getUserReservations.fulfilled, (state, action) => {
        return { ...state, status: 'success', data: action.payload };
      })
      .addCase(getUserReservations.rejected, (state, action) => {
        return { ...state, status: 'error', error: action.error.message };
      });
  },
});

export default reservationSlice.reducer;
