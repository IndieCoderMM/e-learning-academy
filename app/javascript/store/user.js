import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

const getCSRFToken = () => {
  const csrfTag = document.querySelector('meta[name=csrf-token]');
  return csrfTag ? csrfTag.content : '';
};

export const loginUser = createAsyncThunk(
  'user/login',
  async ({ username }) => {
    const res = await fetch('/users/login', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'X-CSRF-Token': getCSRFToken(),
      },
      body: JSON.stringify({ user: { name: username } }),
    });
    const data = await res.json();

    if (!res.ok) {
      throw new Error(data.message);
    }

    return data;
  },
);

export const registerUser = createAsyncThunk(
  'user/register',
  async ({ username }) => {
    const res = await fetch('/users/register', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'X-CSRF-Token': getCSRFToken(),
      },
      body: JSON.stringify({ user: { name: username } }),
    });

    const data = await res.json();

    if (!res.ok) {
      throw new Error(data.message);
    }
    return data;
  },
);

const initialState = {
  status: 'idle',
  name: '',
  id: null,
  error: '',
};

const userSlice = createSlice({
  name: 'user',
  initialState,
  reducers: {
    logOutUser(state) {
      return { ...state, status: 'idle', name: '', id: null };
    },
  },
  extraReducers(builder) {
    builder
      .addCase(registerUser.fulfilled, (state, action) => ({
        ...state,
        status: 'success',
        name: action.payload.name,
        id: action.payload.id,
      }))
      .addCase(registerUser.rejected, (state, action) => ({
        ...state,
        status: 'error',
        error: action.error.message,
      }))
      .addCase(loginUser.fulfilled, (state, action) => ({
        ...state,
        status: 'success',
        name: action.payload.name,
        id: action.payload.id,
      }))
      .addCase(loginUser.rejected, (state, action) => ({
        ...state,
        status: 'error',
        error: action.error.message,
      }));
  },
});

export const { logOutUser } = userSlice.actions;
export default userSlice.reducer;

