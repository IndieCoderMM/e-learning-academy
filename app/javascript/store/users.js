import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

export const logInUser = createAsyncThunk(
  'users/login',
  async ({ username }) => {
    const res = await fetch('/users/login', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ username }),
    });

    const data = await res.json();
    console.log(data);
    return data;
  },
);

export const registerUser = createAsyncThunk(
  'users/register',
  async ({ username }) => {
    const res = await fetch('/users/register', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ username }),
    });
    const data = await res.json();
    console.log(data);
    return data;
  },
);

const initialState = {
  status: 'idle',
  name: '',
  id: null,
  error: '',
  message: '',
};

const userSlice = createSlice({
  name: 'users',
  initialState,
  extraReducers(builder) {
    builder
      .addCase(registerUser.fulfilled, (state, action) => {
        return { ...state, name: action.payload.name, id: action.payload.id };
      })
      .addCase(registerUser.rejected, (state, action) => {
        return { ...state, status: 'error', error: action.payload.errors };
      })
      .addCase(logInUser.fulfilled, (state, action) => {
        return {
          ...state,
          status: 'success',
          name: action.payload.name,
          id: action.payload.id,
        };
      })
      .addCase(logInUser.rejected, (state, action) => {
        return { ...state, status: 'error', error: action.payload.errors };
      });
  },
});

export default userSlice.reducer;
