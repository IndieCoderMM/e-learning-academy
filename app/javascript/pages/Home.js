import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { coursesActions } from '../store/coursesSlice';
import CourseItem from '../components/CourseItem';
import CustomCarousel from '../components/CustomCarousel';

const Home = () => {
  const dispatch = useDispatch();
  const courses = useSelector((state) => state.courses.courses);
  const loading = useSelector((state) => state.courses.loading);
  let coursesItems = [];

  useEffect(() => {
    dispatch(coursesActions.fetchCourses());
  }, [dispatch]);

  if (courses.length) {
    coursesItems = [];
    courses.forEach((course) => {
      const item = {
        element: <CourseItem course={course} />,
        key: course.id,
      };
      coursesItems.push(item);
    });
  }

  if (loading) {
    return <div>Loading...</div>;
  }

  return (
    <section className="page">
      <h2 className="page__title">
        Explore Featured Courses&nbsp;
        <small>
          (<b>{coursesItems.length}</b>
          &nbsp;Available)
        </small>
      </h2>
      <CustomCarousel items={coursesItems} />
    </section>
  );
};

export default Home;
