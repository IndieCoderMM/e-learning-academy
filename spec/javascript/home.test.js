import React from 'react';
import axios from 'axios';
import { render, screen, act } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import store from '../../app/javascript/store/configureStore';
import Home from '../../app/javascript/pages/Home';

jest.mock('../../app/javascript/components/CustomCarousel', () => ({
  __esModule: true,
  default: jest.fn().mockImplementation(({ items }) => (
    <div data-testid="mocked-custom-carousel">
      {items.map((item) => (
        <div key={item.key}>{item.element.props.course.title}</div>
      ))}
    </div>
  )),
}));

jest.mock('axios');

const responseData = [
  {
    id: 1,
    title: 'Course 1',
    description: 'Description for course 1',
    instructor: 'Instructor 1',
    price: 5.99,
    duration: 300,
    img_url: 'image1.jpg',
  },
  {
    id: 2,
    title: 'Course 2',
    description: 'Description for course 2',
    instructor: 'Instructor 2',
    price: 5.99,
    duration: 300,
    img_url: 'image2.jpg',
  },
];

describe('Home page', () => {
  it('displays all available courses', async () => {
    // Arrange
    axios.get.mockResolvedValue({ data: responseData });
    let asFragment;

    // Act
    await act(() => {
      const component = render(
        <BrowserRouter>
          <Provider store={store}>
            <Home />
          </Provider>
          ,
        </BrowserRouter>,
      );
      asFragment = component.asFragment;
    });
    await screen.findByText('Course 1');

    // Assert
    expect(asFragment()).toMatchSnapshot();
    expect(
      screen.getByRole('heading', { name: /explore featured courses/i }),
    ).toBeInTheDocument();
    expect(
      screen.getByRole('heading', { name: /\(\s*2\s*Available\s*\)/i }),
    ).toBeInTheDocument();
    expect(screen.getByText('Course 2')).toBeInTheDocument();
  });
});
