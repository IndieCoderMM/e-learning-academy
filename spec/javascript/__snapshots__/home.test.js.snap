// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home displays all available courses 1`] = `
<DocumentFragment>
  <section
    class="page"
  >
    <h2
      class="page__title"
    >
      Explore Featured Courses 
      <small>
        (
        <b>
          2
        </b>
         Available)
      </small>
    </h2>
    <div
      data-testid="mocked-custom-carousel"
    >
      <div>
        Course 1
      </div>
      <div>
        Course 2
      </div>
    </div>
  </section>
  ,
</DocumentFragment>
`;
